pyccl/pkresponse.py

```diff
@@ -0,0 +1,803 @@
+            )
+
+        _darkemu_set_cosmology(emu, cosmo_hm)
+        for m in range(nM):
+            Pnth_hm[m] = (
+                emu.get_phm(
+                    k_emu * (h / hm), np.log10(nths[m] * (1 / hm) ** 3), z
+                )
+                * (1 / hm) ** 3
+            )
+
+        Nc = prof_hod._Nc(M, aa)
+        Ns = prof_hod._Ns(M, aa)
+        fc = prof_hod._fc(aa)
+        Ng = Nc * (fc + Ns)
+        logMps = logM + dlogM
+        logMms = logM - dlogM
+
+        prof_Mp = prof_hod.fourier(cosmo, k_use, (10**logMps), aa)
+        prof_Mm = prof_hod.fourier(cosmo, k_use, (10**logMms), aa)
+
+        dprof_dlogM = (prof_Mp - prof_Mm) / (2 * dlogM)
+
+        nth_mat = np.tile(nths, (len(k_use), 1)).transpose()
+
+        # Eq. 18
+        ng = hmc._integrate_over_mf(Ng)
+
+        # Eq. 17
+        bgE = hmc._integrate_over_mbf(Ng) / ng
+
+        # Eq. 19
+        bgE2 = hmc._integrate_over_mf(Ng * _b2H17(hmc._bf)) / ng
+
+        bgL = bgE - 1
+
+        b1L_th_mat = np.tile(b1_th_tink - 1, (len(k_emu), 1)).transpose()
+
+        dPhm_db_nfix = (
+            (26.0 / 21.0)
+            * np.log(np.array(Pnth_hp) / np.array(Pnth_hm))
+            * np.array(Pth)
+            / (2 * (np.log(Dp[ia]) - np.log(Dm[ia])))
+        )  # Mpc^3
+
+        dnP_hm_db_emu = nth_mat * (dPhm_db_nfix + b1L_th_mat * np.array(Pbin))
+
+        # Eq. A2
+        nP = nth_mat * np.array(Pth)
+
+        # Eq. A7
+        Pgm = integrate.romb(dprof_dlogM * nP, dx=dlogM, axis=0) / ng
+
+        # The first term of Eq. A8
+        dnP_gm_db = integrate.romb(
+            dprof_dlogM * (dnP_hm_db_emu), dx=dlogM, axis=0
+        )
+
+        # Here we assume the galaxies' profile around host halo
+        # is fixed at physical corrdinate.
+        dprof_dlogM_dlogk = np.zeros((nM, nk))
+        for i in range(nM):
+            dprof_dlogM_dlogk[i] = np.gradient((dprof_dlogM[i])) / np.gradient(
+                np.log(k_use)
+            )
+
+        # The second term of Eq. A8
+        G_prof = (
+            +1.0
+            / 3.0
+            * integrate.romb(dprof_dlogM_dlogk * nP, dx=dlogM, axis=0)
+        )
+
+        # Eq. 25
+        Pgm_growth = (dnP_gm_db + G_prof) / ng - bgL * Pgm
+
+        Pgm_d = (
+            -1.0
+            / 3.0
+            * np.gradient(np.log(Pgm))
+            / np.gradient(np.log(k_use))
+            * Pgm
+        )
+
+        # Eq. 22
+        dPgm_db_emu = Pgm_growth + Pgm_d
+
+        dpklin = pk2dlin(k_use, aa, cosmo, derivative=True)
+
+        # Eq. 16
+        dPgm_db_lin = (
+            (47 / 21 + bgE2 / bgE - bgE - 1 / 3 * dpklin)
+            * bgE
+            * pk2dlin(k_use, aa, cosmo)
+        )
+
+        # stitching
+        k_switch = 0.08  # [h/Mpc]
+
+        # Eq. 27
+        dPgm_db = dPgm_db_lin * np.exp(-k_emu / k_switch) + dPgm_db_emu * (
+            1 - np.exp(-k_emu / k_switch)
+        )
+
+        # use the perturbation theory below kmin
+        kmin = 1e-2  # [h/Mpc]
```

Comment:
I see that you use this in different functions. It would be better if it could be set as a global quantity so that if you change it once, you change it for all

Also, the units of the k and Pk's in CCL are without little `h`. I think you will need to account for this

Reply:
I made `khmin` a global quantity set at the first line of the `pkresponse.py`, while allowing users to set non default value via the argument of the functions. 

I corrected the units of k, and I believe I took into account the units correctly now.

Reply:
Good, thanks!